src/System.Management.Automation/engine/Subsystem/SubsystemInfo.cs

```diff
@@ -0,0 +1,346 @@
+        {
+            _syncObj = new object();
+            _cachedImplInfos = Utils.EmptyReadOnlyCollection<ImplementationInfo>();
+
+            Kind = kind;
+            SubsystemType = subsystemType;
+            AllowUnregistration = false;
+            AllowMultipleRegistration = false;
+            RequiredCmdlets = Utils.EmptyReadOnlyCollection<string>();
+            RequiredFunctions = Utils.EmptyReadOnlyCollection<string>();
+        }
+
+        private protected abstract void AddImplementation(ISubsystem rawImpl);
+        private protected abstract ISubsystem RemoveImplementation(Guid id);
+
+        internal void RegisterImplementation(ISubsystem impl)
+        {
+            AddImplementation(impl);
+        }
+
+        internal ISubsystem UnregisterImplementation(Guid id)
+        {
+            return RemoveImplementation(id);
+        }
+
+        #endregion
+
+        #region "Static factory overloads"
+
+        internal static SubsystemInfo Create<TConcreteSubsystem>(SubsystemKind kind)
+            where TConcreteSubsystem : class, ISubsystem
+        {
+            return new SubsystemInfoImpl<TConcreteSubsystem>(kind);
+        }
+
+        internal static SubsystemInfo Create<TConcreteSubsystem>(
+            SubsystemKind kind,
+            bool allowUnregistration,
+            bool allowMultipleRegistration) where TConcreteSubsystem : class, ISubsystem
+        {
+            return new SubsystemInfoImpl<TConcreteSubsystem>(kind)
+            {
+                AllowUnregistration = allowUnregistration,
+                AllowMultipleRegistration = allowMultipleRegistration,
+            };
+        }
+
+        internal static SubsystemInfo Create<TConcreteSubsystem>(
+            SubsystemKind kind,
+            bool allowUnregistration,
+            bool allowMultipleRegistration,
+            ReadOnlyCollection<string> requiredCmdlets,
+            ReadOnlyCollection<string> requiredFunctions) where TConcreteSubsystem : class, ISubsystem
+        {
+            if (allowMultipleRegistration &&
+                (requiredCmdlets.Count > 0 || requiredFunctions.Count > 0))
+            {
+                throw new ArgumentException(
+                    StringUtil.Format(
+                        SubsystemStrings.InvalidSubsystemInfo,
+                        kind.ToString()));
+            }
+
+            return new SubsystemInfoImpl<TConcreteSubsystem>(kind)
+            {
+                AllowUnregistration = allowUnregistration,
+                AllowMultipleRegistration = allowMultipleRegistration,
+                RequiredCmdlets = requiredCmdlets,
+                RequiredFunctions = requiredFunctions,
+            };
+        }
+
+        #endregion
+
+        #region "ImplementationInfo"
+
+        /// <summary>
+        /// Information about an implementation of a subsystem.
+        /// </summary>
+        public class ImplementationInfo
```

Comment:
Can it be a readonly struct?

Reply:
I don't see a reason to do this. There won't be any benefit in allocation because there is 1 such instance corresponding to 1 subsystem implementation, so the total number will be very limited.